### ISSUE TITLE
combineCandidates will sometimes fail and cause assertion failure during network unit tests

### ISSUE THREAD

**Issue by hewison-chris** [bosagora/agora#1250]
Occasionally when a network test executes it fails with an assertion in the combineCandidates function in the Nominator with error similar to 
#### FATAL ERROR: [source/agora/consensus/protocol/Nominator.d:705] combineCandidates: no valid candidate for slot 16.

Maybe this indicates a potential problem in production.

**Comment by hewison-chris** [comment]
With some extra logging:-
```
public override Value combineCandidates (uint64_t slot_idx,
        ref const(set!Value) candidates)
    {
        try
        {
            foreach (ref const(Value) candidate; candidates)
            {
                auto data = deserializeFull!ConsensusData(candidate[]);

                if (auto msg = this.ledger.validateConsensusData(data))
                {
                    log.error("combineCandidates(): Invalid consensus data: {}", msg);
                    continue;
                }
                else
                {
                    log.info("combineCandidates: {}", slot_idx);
                }

                // todo: currently we just pick the first of the candidate values,
                // but we should ideally pick tx's out of the combined set
                return duplicate_value(&candidate);
            }
        } catch (Exception ex)
        {
            assert(0, format!"[%s:%s] combineCandidates: slot %u. Exception: %s"(
                __FILE__, __LINE__, slot_idx, ex.to!string));
        }
        import std.stdio : writefln;
        try
        {
            writefln("SHOULD NOT BE HERE!!!! candidates=");
            foreach (ref const(Value) candidate; candidates)
            {
                writefln("consesusData=%s", deserializeFull!ConsensusData(candidate[]));
            }
        }
        catch (Exception) {}
        assert(0, format!"[%s:%s] combineCandidates: no valid candidate for slot %u. Candidates:%s"(
            __FILE__, __LINE__, slot_idx, candidates));
    }

SHOULD NOT BE HERE!!!! candidates=
consesusData=ConsensusData([Transaction(Payment, [Input(0x4b87e702b639bf31ed9fc87fa2812eb4cbf2d945eacf84eee5175a3690a730dff427bbc7baa284f9e061ec79c0e493eab70fc51cd3ed15d7030bfe15cd0296a1, 0, 0x06386763e000d70df43b7751b5b9d4c5d7c5e70e0a0e3ff1232c63d55f574fef58c10cdf140ec6ab3b3074f362c321eeeeb3c8e5985c1dca82005f6dd513fbde)], [Output(610000000000000, GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ)])], [])
#### FATAL ERROR: [source/agora/consensus/protocol/Nominator.d:714] combineCandidates: no valid candidate for slot 13. Candidates:const(set!(xvector!(ubyte, 4294967292u)))([7F9BF58B3660, 7F9BF58B3660, 1])
This node was started at source/agora/test/ValidatorCleanRestart.d:39
This most likely means that the node crashed due to an uncaught exception
If not, please file a bug at https://github.com/Geod24/localrest/
Full error: core.exception.AssertError@source/agora/consensus/protocol/Nominator.d(713): [source/agora/consensus/protocol/Nominator.d:714] combineCandidates: no valid candidate for slot 13. Candidates:const(set!(xvector!(ubyte, 4294967292u)))([7F9BF58B3660, 7F9BF58B3660, 1])
----------------
??:? onAssertErrorMsg [0x10a8f1394]
??:? _d_assert_msg [0x10a8f17c5]
??:? _ZN9Nominator17combineCandidatesEyRKSt3setIN3xdr7xvectorIhLj4294967292EEEE [0x109db4555]
??:? _ZN7stellar18NominationProtocol15processEnvelopeERKNS_11SCPEnvelopeE [0x10a33701e]
??:? _ZN7stellar4Slot15processEnvelopeERKNS_11SCPEnvelopeEb [0x10a38b783]
??:? _ZN7stellar18NominationProtocol14emitNominationEv [0x10a33512f]
??:? _ZN7stellar18NominationProtocol8nominateERKN3xdr7xvectorIhLj4294967292EEES5_b [0x10a337e4d]
??:? _ZN7stellar4Slot8nominateERKN3xdr7xvectorIhLj4294967292EEES5_b [0x10a38c51c]
??:? _ZN7stellar3SCP8nominateEyRKN3xdr7xvectorIhLj4294967292EEES5_ [0x10a37b06c]
??:? @trusted void agora.consensus.protocol.Nominator.Nominator.nominate(ulong, agora.consensus.data.ConsensusData.ConsensusData) [0x109db5169]
??:? @safe void agora.consensus.protocol.Nominator.Nominator.checkNominate() [0x109e66288]
??:? void geod24.LocalRest.Timer.run() [0x10a42ff36]
??:? void geod24.concurrency.FiberScheduler.create(void delegate()).wrap() [0x10a431ed8]
??:? void core.thread.context.Callable.opCall() [0x10a8fe9ce]
??:? void core.thread.fiber.Fiber.run() [0x10a8feb5d]
??:? fiber_entryPoint [0x10a8fea92]

**Comment by AndrejMitrovic** [comment]
The question is: is that code path reached because the candidate set was empty, or because it failed validation?

If it failed validation it would be good to have some extra information. For example, what is the current ledger block height, and also what is the slot index we are currently nominating.

**Comment by AndrejMitrovic** [comment]
Can you reliably replicate the bug? Can you narrow it down to a specific test? E.g. `dtest=agora.test.Quorum dub test`

**Comment by hewison-chris** [comment]
Unfortunately it does not happen all the time. On the same test it can occur at different heights and then run fine.

**Comment by AndrejMitrovic** [comment]
I'm guessing this only happens with your new changes? Not in `v0.x.x`? I haven't seen this bug yet in `v0.x.x`.

**Comment by hewison-chris** [comment]
I did just catch it again and it is due to validateConsensusData returning something. However the reason is not logged as the assert throws too early for this to happen.

**Comment by hewison-chris** [comment]
I will add some more temporary logging to see the reason for it being invalid.

**Comment by AndrejMitrovic** [comment]
I can think of two reasons why `validateConsensusData` could fail in `combineCandidates`:
- Memory corruption. The passed in candidate set is somehow corrupted
- The candidate set is stale. Perhaps there is a race condition where externalize was called but a fiber switched and then the ~ballot~ nominating timer fired again in the meantime

I suspect it could be the second case..

**Comment by AndrejMitrovic** [comment]
Sorry I meant the Nominating timer, not balloting. It's the nominating timer which calls `combineCandidates`.

**Comment by hewison-chris** [comment]
`combineCandidates: Invalid consensus data: Transaction: Input ref not in UTXO`
Seems that this is the reason. However I don't think that should cause an assertion failure in the node? It should be rejected.

**Comment by AndrejMitrovic** [comment]
No this is definitely a real bug somewhere. `combineCandidates` takes a series of **validated** candidates and creates a single composite candidate. We currently just pick the first one, but this part is left to be done later once we implement transaction filtering.

All of the candidate values should be valid.

**Comment by AndrejMitrovic** [comment]
> combineCandidates: Invalid consensus data: Transaction: Input ref not in UTXO

Can you also print out the current block height, and the `slot_idx`?

**Comment by hewison-chris** [comment]
combineCandidates: Invalid consensus data: Transaction: Input ref not in UTXO
SHOULD NOT BE HERE!!!! candidates=
consesusData=ConsensusData([Transaction(Payment, [Input(0xf5947c767253f1312422a7dafbaba9c0469fd3f8fb3b760d2c0c849f11fcec326144d82168496369f2c221072b5bd079b2408f8a566f0659696589189729a8fe, 0, 0x0eea252a8e34a180b134aca58108cf46e4b32e477129e8a51bedfdd149eb7c7dda6528174928703dea8a333f4de3dbd12b7c320537ce65811c10489171686089)], [Output(610000000000000, GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ)])], [])
#### FATAL ERROR: [source/agora/consensus/protocol/Nominator.d:716] combineCandidates: no valid candidate for slot 5. Candidates:const(set!(xvector!(ubyte, 4294967292u)))([7FB9CEC454B0, 7FB9CEC454B0, 1])
This node was started at source/agora/test/ValidatorCleanRestart.d:39
This most likely means that the node crashed due to an uncaught exception
If not, please file a bug at https://github.com/Geod24/localrest/
Full error: core.exception.AssertError@source/agora/consensus/protocol/Nominator.d(715): [source/agora/consensus/protocol/Nominator.d:716] combineCandidates: no valid candidate for slot 5. Candidates:const(set!(xvector!(ubyte, 4294967292u)))([7FB9CEC454B0, 7FB9CEC454B0, 1])
----------------
??:? onAssertErrorMsg [0x1109ee354]
??:? _d_assert_msg [0x1109ee785]
??:? _ZN9Nominator17combineCandidatesEyRKSt3setIN3xdr7xvectorIhLj4294967292EEEE [0x10feb1505]
??:? _ZN7stellar18NominationProtocol15processEnvelopeERKNS_11SCPEnvelopeE [0x1104340fe]
??:? _ZN7stellar4Slot15processEnvelopeERKNS_11SCPEnvelopeEb [0x110488863]
??:? _ZN7stellar18NominationProtocol14emitNominationEv [0x11043220f]
??:? _ZN7stellar18NominationProtocol8nominateERKN3xdr7xvectorIhLj4294967292EEES5_b [0x110434f2d]
??:? _ZN7stellar4Slot8nominateERKN3xdr7xvectorIhLj4294967292EEES5_b [0x1104895fc]
??:? _ZN7stellar3SCP8nominateEyRKN3xdr7xvectorIhLj4294967292EEES5_ [0x11047814c]
??:? @trusted void agora.consensus.protocol.Nominator.Nominator.nominate(ulong, agora.consensus.data.ConsensusData.ConsensusData) [0x10feb2119]
??:? @safe void agora.consensus.protocol.Nominator.Nominator.checkNominate() [0x10ff63238]
??:? void geod24.LocalRest.Timer.run() [0x11052d016]
??:? void geod24.concurrency.FiberScheduler.create(void delegate()).wrap() [0x11052efb8]
??:? void core.thread.context.Callable.opCall() [0x1109fb98e]
??:? void core.thread.fiber.Fiber.run() [0x1109fbb1d]
??:? fiber_entryPoint [0x1109fba52]
source/agora/test/ValidatorCleanRestart.d(42): Node logs:

**Comment by AndrejMitrovic** [comment]
Can you check the block height too? It should be `this.ledger.getBlockHeight()`

**Comment by hewison-chris** [comment]
> Can you check the block height too? It should be `this.ledger.getBlockHeight()`

Next time it occurs I will have it logged also.

**Comment by hewison-chris** [comment]
combineCandidates: Invalid consensus data: Transaction: Input ref not in UTXO
SHOULD NOT BE HERE!!!! candidates=
Block Height=5
consesusData=ConsensusData([Transaction(Payment, [Input(0xf5947c767253f1312422a7dafbaba9c0469fd3f8fb3b760d2c0c849f11fcec326144d82168496369f2c221072b5bd079b2408f8a566f0659696589189729a8fe, 0, 0x0eea252a8e34a180b134aca58108cf46e4b32e477129e8a51bedfdd149eb7c7dda6528174928703dea8a333f4de3dbd12b7c320537ce65811c10489171686089)], [Output(610000000000000, GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ)])], [])
#### FATAL ERROR: [source/agora/consensus/protocol/Nominator.d:717] combineCandidates: no valid candidate for slot 5. Candidates:const(set!(xvector!(ubyte, 4294967292u)))([7FDF31828B10, 7FDF31828B10, 1])

**Comment by hewison-chris** [comment]
Looks like it was for last block.

**Comment by AndrejMitrovic** [comment]
Yes this is what I feared. It looks like the nomination protocol has continued for the same slot **after** externalization. I'll look at the code to see what could trigger this.

**Comment by hewison-chris** [comment]
I'll push my current code to a different branch as it is quite easy to reproduce to the moment.

**Comment by hewison-chris** [comment]
This is branch https://github.com/hewison-chris/agora/tree/combineCandidatesBug
Test to trigger quite often is `dtest=agora.test.ValidatorCleanRestart dub test`

**Comment by AndrejMitrovic** [comment]
I thought I already covered this case though:

https://github.com/hewison-chris/agora/blob/c1d52be671b0cbdb9b4c01f17321b1d3341595b3/source/agora/node/Validator.d#L341-L343

But maybe the race condition is more subtle..

**Comment by AndrejMitrovic** [comment]
I wonder if there is just a missing call to `stopNominationRound` inside of the `valueExternalized` routine..

**Comment by AndrejMitrovic** [comment]
Hmm no, wrong assumption. It is eventually called in https://github.com/hewison-chris/agora/blob/c1d52be671b0cbdb9b4c01f17321b1d3341595b3/source/agora/node/Validator.d#L341-L343.

**Comment by hewison-chris** [comment]
I can no longer reproduce it when I add a check that we are not externalizing a block during prep of consensusData:-
```
 public void prepareNominatingSet (ref ConsensusData data, ulong max_txs)
        @safe
    {
        if (externalizingBlock) return;
```
and
```
public bool onExternalized (ConsensusData data)
        @trusted
    {
        scope(exit) externalizingBlock = false;
        externalizingBlock = true;
        auto block = makeNewBlock(this.last_block, data.tx_set, data.enrolls);
        return this.acceptBlock(block);
    }
```

**Comment by AndrejMitrovic** [comment]
Hmm.. that would mean that `checkNominate` is called before `onExternalized` finished running. I think that means there is a context switch somewhere.

**Comment by AndrejMitrovic** [comment]
I can see another source of a context switch:

https://github.com/bpfkorea/agora/blob/d3b05dcea8414882180e5b73014f468c346974af/source/scpp/src/scp/NominationProtocol.cpp#L426-L435

`emitNomination()` ultimately calls `Nominator.emitEnvelope()`, which calls `sendEnvelope` on each client via `runTask`. And `runTask` yields the calling fiber.

The problem with yielding the fiber is that the ledger state can change in the meantime.

We really need a `schedule()` primitive here.

**Comment by AndrejMitrovic** [comment]
This diff essentially uses schedule under the hood (at least in LocalRest):

```diff
diff --git a/source/agora/network/NetworkClient.d b/source/agora/network/NetworkClient.d
index 45cbc4935..22d7bb399 100644
--- a/source/agora/network/NetworkClient.d
+++ b/source/agora/network/NetworkClient.d
@@ -239,7 +239,7 @@ class NetworkClient

     public void sendEnvelope (SCPEnvelope envelope) nothrow
     {
-        this.taskman.runTask(
+        this.taskman.setTimer(0.seconds,
         {
             this.attemptRequest!(API.receiveEnvelope, Throw.No)(this.api,
                 envelope);
```

I think it fixes the `combineCandidates` bug. But there might be other bugs there still because the tests in your branch still don't succeed for me.

**Comment by hewison-chris** [comment]
> This diff essentially uses schedule under the hood (at least in LocalRest):
> 
> ```diff
> diff --git a/source/agora/network/NetworkClient.d b/source/agora/network/NetworkClient.d
> index 45cbc4935..22d7bb399 100644
> --- a/source/agora/network/NetworkClient.d
> +++ b/source/agora/network/NetworkClient.d
> @@ -239,7 +239,7 @@ class NetworkClient
> 
>      public void sendEnvelope (SCPEnvelope envelope) nothrow
>      {
> -        this.taskman.runTask(
> +        this.taskman.setTimer(0.seconds,
>          {
>              this.attemptRequest!(API.receiveEnvelope, Throw.No)(this.api,
>                  envelope);
> ```
> 
> I think it fixes the `combineCandidates` bug. But there might be other bugs there still because the tests in your branch still don't succeed for me.

Thanks @AndrejMitrovic I'll try this out. The tests are still being refactored so I would not expect them to pass yet.

**Comment by Geod24** [comment]
Fixed by #1256
